core/src/main/scala/org/apache/gearpump/util/Graph.scala

```diff
@@ -81,28 +86,35 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
    * out degree
    */
   def outDegreeOf(node: N): Int = {
-    edges.count(_._1 == node)
+    outgoingEdgesOf(node).size
   }
 
   /**
    * in degree
    */
   def inDegreeOf(node: N): Int = {
-    edges.count(_._3 == node)
+    incomingEdgesOf(node).size
   }
 
   /**
    * out going edges.
    */
   def outgoingEdgesOf(node: N): List[(N, E, N)] = {
-    edges.filter(_._1 == node)
+    _outEdges.getOrElse(node, List.empty)
   }
 
   /**
    * incoming edges.
    */
   def incomingEdgesOf(node: N): List[(N, E, N)] = {
-    edges.filter(_._3 == node)
+    _inEdges.getOrElse(node, List.empty)
+  }
+
+  /**
+   * adjacent vertices.
+   */
+  def adjacentVertices(node: N): List[N] = {
```

Comment:
how about vertices of incoming edges ?

Reply:
It's a directed graph

Reply:
can this be private ? We don't know this is a DAG from outside.

Reply:
sure